File: code/fts_fuzzy_match.cs
// LICENSE
//
//   This software is dual-licensed to the public domain and under the following
//   license: you are granted a perpetual, irrevocable license to copy, modify,
//   publish, and distribute this file as you see fit.

// VERSION 0.1.0

using System;
using System.Collections.Generic;

namespace fts
{
	public static class FuzzyMatcher
	{
		/// <summary>
		/// Does a fuzzy search for a pattern within a string.
		/// </summary>
		/// <param name="pattern">The pattern to search for in the string.</param>
		/// <param name="str">The string to search for the pattern in.</param>
		/// <returns>true if each character in pattern is found sequentially within str; otherwise, false.</returns>
		public static bool FuzzyMatch(string pattern, string str)
		{
			var patternIdx = 0;
			var strIdx = 0;
			var patternLength = pattern.Length;
			var strLength = str.Length;

			while (patternIdx != patternLength && strIdx != strLength)
			{
				if (char.ToLower(pattern[patternIdx]) == char.ToLower(str[strIdx]))
					++patternIdx;
				++strIdx;
			}

			return patternLength != 0 && strLength != 0 && patternIdx == patternLength;
		}

		/// <summary>
		/// Does a fuzzy search for a pattern within a string, and gives the search a score on how well it matched.
		/// </summary>
		/// <param name="pattern">The pattern to search for in the string.</param>
		/// <param name="str">The string to search for the pattern in.</param>
		/// <param name="outScore">The score which this search received, if a match was found.</param>
		/// <returns>true if each character in pattern is found sequentially within str; otherwise, false.</returns>
		public static bool FuzzyMatch(string pattern, string str, out int outScore)
		{
			// Score consts
			const int adjacencyBonus = 5;              // bonus for adjacent matches
			const int separatorBonus = 10;             // bonus if match occurs after a separator
			const int camelBonus = 10;                 // bonus if match is uppercase and prev is lower

			const int leadingLetterPenalty = -3;       // penalty applied for every letter in str before the first match
			const int maxLeadingLetterPenalty = -9;    // maximum penalty for leading letters
			const int unmatchedLetterPenalty = -1;     // penalty for every letter that doesn't matter


			// Loop variables
			var score = 0;
			var patternIdx = 0;
			var patternLength = pattern.Length;
			var strIdx = 0;
			var strLength = str.Length;
			var prevMatched = false;
			var prevLower = false;
			var prevSeparator = true;                  // true if first letter match gets separator bonus

			// Use "best" matched letter if multiple string letters match the pattern
			char? bestLetter = null;
			char? bestLower = null;
			int? bestLetterIdx = null;
			var bestLetterScore = 0;

			var matchedIndices = new List<int>();

			// Loop over strings
			while (strIdx != strLength)
			{
				var patternChar = patternIdx != patternLength ? pattern[patternIdx] as char? : null;
				var strChar = str[strIdx];

				var patternLower = patternChar != null ? char.ToLower((char)patternChar) as char? : null;
				var strLower = char.ToLower(strChar);
				var strUpper = char.ToUpper(strChar);

				var nextMatch = patternChar != null && patternLower == strLower;
				var rematch = bestLetter != null && bestLower == strLower;

				var advanced = nextMatch && bestLetter != null;
				var patternRepeat = bestLetter != null && patternChar != null && bestLower == patternLower;
				if (advanced || patternRepeat)
				{
					score += bestLetterScore;
					matchedIndices.Add((int)bestLetterIdx);
					bestLetter = null;
					bestLower = null;
					bestLetterIdx = null;
					bestLetterScore = 0;
				}

				if (nextMatch || rematch)
				{
					var newScore = 0;

					// Apply penalty for each letter before the first pattern match
					// Note: Math.Max because penalties are negative values. So max is smallest penalty.
					if (patternIdx == 0)
					{
						var penalty = Math.Max(strIdx * leadingLetterPenalty, maxLeadingLetterPenalty);
						score += penalty;
					}

					// Apply bonus for consecutive bonuses
					if (prevMatched)
						newScore += adjacencyBonus;

					// Apply bonus for matches after a separator
					if (prevSeparator)
						newScore += separatorBonus;

					// Apply bonus across camel case boundaries. Includes "clever" isLetter check.
					if (prevLower && strChar == strUpper && strLower != strUpper)
						newScore += camelBonus;

					// Update pattern index IF the next pattern letter was matched
					if (nextMatch)
						++patternIdx;

					// Update best letter in str which may be for a "next" letter or a "rematch"
					if (newScore >= bestLetterScore)
					{
						// Apply penalty for now skipped letter
						if (bestLetter != null)
							score += unmatchedLetterPenalty;

						bestLetter = strChar;
						bestLower = char.ToLower((char)bestLetter);
						bestLetterIdx = strIdx;
						bestLetterScore = newScore;
					}

					prevMatched = true;
				}
				else
				{
					score += unmatchedLetterPenalty;
					prevMatched = false;
				}

				// Includes "clever" isLetter check.
				prevLower = strChar == strLower && strLower != strUpper;
				prevSeparator = strChar == '_' || strChar == ' ';

				++strIdx;
			}

			// Apply score for last match
			if (bestLetter != null)
			{
				score += bestLetterScore;
				matchedIndices.Add((int)bestLetterIdx);
			}

			outScore = score;
			return patternIdx == patternLength;
		}
	}
}

File: code/fts_fuzzy_match.php
<?php
function fuzzy_match($pattern, $str) {
	//score constants
	$ADJACENCY_BONUS = 5; //bonus for adjacent matches
	$SEPERATOR_BOUNUS = 10; //bonus if match occures after a separator
	$CAMEL_BONUS = 10; // bonus if match is uppercase and prev is lower
	$LEADING_LETTER_PENALTY = -3; //penalty applied for evert letter in the $str before the first match
	$MAX_LEADING_LETTER_PENALTY = -9; //max penality for leading letters
	$UNMATCHED_LETTER_PENALTY = -1; // penalty for evert letter that doesn't matter

	//loop variables
	$score = 0;
	$patternIdx = 0;
	$patternLength = strlen($pattern);
	$strIdx = 0;
	$strLength = strlen(str);
	$prevMatched = false;
	$prevLower = false;
	$prevSeparator = true; //true so if the first letter match gets separator bonus

	//use "best" matches letter if multiple string letters match the $pattern
	$bestLetter = null;
	$bestLower = null;
	$bestLetterIdx = null;
	$bestLetterScore = 0;

	$matchedIndecies = [];

	while ($strIdx != $strLength) {
		$patternChar = ($patternIdx != $patternLength) ? $pattern{$strIdx} : null; //check char at current position of index
		$strChar = $str{$strIdx}; //current letter check for entered string
		$patternLower = ($patternChar != null) ? strtolower($patternChar) : null; //converts pattern to lower
		
		//str conversion to upper/lower
		$strLower = strtolower($str);
		$strUpper = strtoupper($str);

		$nextMatch = ($patternChar && $patternLower == $strLower);
		$rematch = ($bestletteer && $patternChar && $bestLower == $strLower);
		$advanced = ($nextMatch && $bestLetter);
		$patternRepeat = ($bestLetter && $patternChar && $bestLower == $patternLower);

		//push to $matchedIndecies if $patternChar and $patternLower are equal to $strLower 
		//		->(so a matched char between pattern and str), 
		//OR pushes to $matchedIndecies if $bestLetter, $patternChar and $bestLower are the same as $patternLower
		//		->(the best letter that matches for lowercase checks)
		if ($advanced || $patternRepeat) {
			$score += $bestLetterScore;
			array_push($matchedIndecies, $bestLetterIdx);
			$bestLetter = null;
			$bestLower= null;
			$bestletterIdx = 0;
			$bestLetterScore = 0;
		}

		if ($nextMatch || $rematch) {
			$newScore = 0;

			//apply penalty for each letter before the first pattern match
			//note: std::max because penalities are negative values. so max is smallest penalty
			if ($pattenIdx == 0) {
				$penalty = max($strIdx * $LEADING_LETTER_PENALTY, $MAX_LEADING_LETTER_PENALTY);
				$score += $penalty;
			}

			//apply bonus to consecutive bonuses
			if($prevMatched) {
				$newScore += $ADJACENCY_BONUS;
			}

			//apply bonus across camel case boundaries. includes "clever" isLetter check
			if ($prevLower && $strChar == $strUpper && $strlower != $strUpper) {
				$newScore += $CAMEL_BONUS;
			}

			//update best letter in str which may be for a "next" letter or a "rematch"
			if ($newScore >= $bestLetterScore) {
				//apply penality for now skipped letter
				if ($bestLetter != null) {
					$score += $UNMATCHED_LETTER_PENALTY;
				}

				$bestletter = $strChar;
				$bestLower = strtolower($bestLetter);
				$bestLetterIdx = $strIdx;
				$bestLetterScore = $newScore;
			}

			$prevMatched = true;

		} else {
			//append unmatch charaters
			$formattedStr += $strChar;

			$score += $UNMATCHED_LETTER_PENALTY;
			$prevMatched = false;
		}

		//includes "clever" isLetter check
		$prevLower = $strChar == $strLower && $strLower != $strUpper;
		$prevSeparator = $strChar == '_' || $strChar == ' ';

		++ $strIdx;
	}

//apply score for last match
	if ($bestLetter) {
		$score += $bestLetterScore;
		array_push($matchedIndecies, $bestLetterIdx);
	}

	//finish out formatter string after last pattern matched
	//build formated string based on matched letters
	$formattedStr = "";
	$lastIdx = 0;
	for ($i = 0; $i < length($matchedIndecies); ++$i) {
		$idx = $matchedIndecies[$i];
		$formattedStr .= substr($str, ($idx - $lastIdx)) . "<b>" . $str{$idx} . "</b>";
		$lastIdx = $idx + 1;
	}
	$formattedStr .= substr($str, $lastIdx, (strlen($str) - $lastIdx));

	$matched = $patternIdx == $patternLength;
	return [$matched, $score, $formattedStr];
}


